### PR TITLE
OAuth 2.0 Login

### PR DESCRIPTION
**Pull request by aditya1702**
This takes place in three steps.
Step-1: The user clicks 'Login with Google' button:
screenshot from 2016-05-12 11 07 47

Step-2: The user is redirected to the google account page:
step 2

Step--3: The user ,if he clicks allow, will be redirected to the admin page:
step 3

### CODE REVIEW

File: open_event/views/views.py [hunk 1]
```diff
@@ -279,28 +292,97 @@ def generate_icalender_event(event_id):
...
+        return redirect(url_for('admin.index'))
+    elif 'error' in request.args:
+        if request.args.get('error') == 'access denied':
+            return 'You denied access'
+        return 'Error encountered'
+    elif 'code' not in request.args and 'state' not in request.args:
+        print "HELLO"
+        return redirect(url_for('admin.login_view'))
+    else:
+        google = get_google_auth()
```
Comment (juslee) [inline open_event/views/views.py]: Codacy Issue found: [Unused variable 'auth_url'](https://www.codacy.com/app/triplez/open-event-orga-server/file/2576208189/issues/source?bid=3233864&fileBranchId=3299682#l326)

File: open_event/views/views.py [hunk 2]
```diff
@@ -279,28 +292,95 @@ def generate_icalender_event(event_id):
...
+        try:
+            token = facebook.fetch_token(FbOAuth.get_token_uri(), authorization_url=request.url,
+                                         code=code_url, client_secret=FbOAuth.get_client_secret())
+        except HTTPError:
+            return 'HTTP Error occurred'
+        facebook = get_facebook_auth(token=token)
+        response = facebook.get(FbOAuth.get_user_info())
+        if resp.status_code == 200:
+            user_info = response.json()
+            email = user_info['email']
```
Comment (juslee) [inline open_event/views/views.py]: Codacy Issue found: [Invalid variable name "u"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2576260486/issues/source?bid=3233864&fileBranchId=3299682#l370)

**Comment by aditya1702** [comment]
Writing tests for this!

File: tests/test_google_oauth.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import unittest
+from tests.utils import OpenEventTestCase
+from flask import url_for
+from tests.setup_database import Setup
+from open_event import current_app as app
+from open_event.models import db
+from open_event.helpers.data import save_to_db, get_google_auth
+from open_event.models.user import User
+from open_event.helpers.oauth import OAuth
+from tests.object_mother import ObjectMother
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused ObjectMother imported from tests.object_mother](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581110445/issues/source?bid=3233864&fileBranchId=3299682#l10)

File: tests/test_google_oauth.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import unittest
+from tests.utils import OpenEventTestCase
+from flask import url_for
+from tests.setup_database import Setup
+from open_event import current_app as app
+from open_event.models import db
+from open_event.helpers.data import save_to_db, get_google_auth
+from open_event.models.user import User
+from open_event.helpers.oauth import OAuth
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused OAuth imported from open_event.helpers.oauth](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581110445/issues/source?bid=3233864&fileBranchId=3299682#l9)

File: tests/test_google_oauth.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import unittest
+from tests.utils import OpenEventTestCase
+from flask import url_for
+from tests.setup_database import Setup
+from open_event import current_app as app
+from open_event.models import db
+from open_event.helpers.data import save_to_db, get_google_auth
+from open_event.models.user import User
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused User imported from open_event.models.user](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581110445/issues/source?bid=3233864&fileBranchId=3299682#l8)

File: tests/test_google_oauth.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import unittest
+from tests.utils import OpenEventTestCase
+from flask import url_for
+from tests.setup_database import Setup
+from open_event import current_app as app
+from open_event.models import db
+from open_event.helpers.data import save_to_db, get_google_auth
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused save_to_db imported from open_event.helpers.data](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581110445/issues/source?bid=3233864&fileBranchId=3299682#l7)

File: tests/test_google_oauth.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import unittest
+from tests.utils import OpenEventTestCase
+from flask import url_for
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused url_for imported from flask](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581110445/issues/source?bid=3233864&fileBranchId=3299682#l3)

File: tests/test_google_oauth.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+from tests.setup_database import Setup
+from open_event import current_app as app
+from open_event.models import db
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused db imported from open_event.models](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581110445/issues/source?bid=3233864&fileBranchId=3299682#l6)

File: tests/test_google_oauth.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import unittest
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused import unittest](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581110445/issues/source?bid=3233864&fileBranchId=3299682#l1)

File: tests/test_google_oauth.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+from tests.utils import OpenEventTestCase
+from flask import url_for
+from tests.setup_database import Setup
+from open_event import current_app as app
+from open_event.models import db
+from open_event.helpers.data import save_to_db, get_google_auth
+from open_event.models.user import User
+from open_event.helpers.oauth import OAuth
+from tests.object_mother import ObjectMother
+from tests.auth_helper import register
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused register imported from tests.auth_helper](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581110445/issues/source?bid=3233864&fileBranchId=3299682#l11)

File: tests/test_google_oauth.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import unittest
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused import unittest](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581443479/issues/source?bid=3233864&fileBranchId=3299682#l1)

Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused import unittest](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581464956/issues/source?bid=3233864&fileBranchId=3299682#l1)

Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused import unittest](https://www.codacy.com/app/triplez/open-event-orga-server/file/2581472215/issues/source?bid=3233864&fileBranchId=3299682#l1)

**Comment by aditya1702** [comment]
@leto How do I write test for this. The API call is made to the Google server. I am not getting exactly how do I test this?

**Comment by aditya1702** [comment]
@mariobehling please review

File: tests/test_google_oauth.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+
+class TestGoogleOauth(OpenEventTestCase):
+    def setUp(self):
+        self.app = Setup.create_app()
+
+    def test_redirect(self):
+        client = WebApplicationClient('some_client_id')
+        uri = "https://gCallback/?state=dummy_state&code=dummy_code/"
+        self.assertTrue('code' in client.parse_request_uri_response(uri, state='dummy_state'))
```
Comment (leto) [inline tests/test_google_oauth.py]: Better to assert the actual value of `code`

File: tests/test_google_oauth.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+class TestGoogleOauth(OpenEventTestCase):
+    def setUp(self):
+        self.app = Setup.create_app()
+
+    def test_redirect(self):
+        """Tests whether on redirection the user is being redirected to the proper authentication url of Google"""
+        with app.test_request_context():
+            google = get_google_auth()
+            auth_url, state = google.authorization_url(OAuth.get_auth_uri(), access_type='offline')
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused variable 'state'](https://www.codacy.com/app/triplez/open-event-orga-server/file/2604774180/issues/source?bid=3233864&fileBranchId=3299682#l18)

File: tests/test_google_oauth.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
+import unittest
+from tests.utils import OpenEventTestCase
+from tests.setup_database import Setup
+from open_event import current_app as app
+from oauthlib.oauth2 import WebApplicationClient
+from open_event.helpers.oauth import OAuth
+from open_event.helpers.data import get_google_auth
+from tests.auth_helper import register, login
+from open_event.models.user import User
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [Unused User imported from open_event.models.user](https://www.codacy.com/app/triplez/open-event-orga-server/file/2605468172/issues/source?bid=3233864&fileBranchId=3299682#l9)

File: migrations/env.py [hunk 15]
```diff
@@ -4,25 +4,26 @@
...
-from flask import current_app
-
 # this is the Alembic Config object, which provides
 # access to the values within the .ini file in use.
 config = context.config
 
 # Interpret the config file for Python logging.
 # This line sets up loggers basically.
 fileConfig(config.config_file_name)
-LOGGER = logging.getLogger('alembic.env')
```
Comment (juslee) [inline migrations/env.py]: Codacy Issue found: [Invalid constant name "logger"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2611506661/issues/source?bid=3233864&fileBranchId=3299682#l14)

File: open_event/views/views.py [hunk 16]
```diff
@@ -1,7 +1,10 @@
 """Copyright 2015 Rafal Kowalski"""
 import os
-from flask import jsonify, url_for, redirect, request, send_from_directory
+import icalendar
+
+from flask import jsonify, url_for, redirect, request, send_from_directory, session
```
Comment (juslee) [inline open_event/views/views.py]: Codacy Issue found: [Unused session imported from flask](https://www.codacy.com/app/triplez/open-event-orga-server/file/2611506722/issues/source?bid=3233864&fileBranchId=3299682#l5)

File: migrations/env.py (hunk 15 above)
Comment (juslee) [inline migrations/env.py]: Codacy Issue found: [Invalid constant name "logger"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2611527092/issues/source?bid=3233864&fileBranchId=3299682#l14)

Comment (juslee) [inline migrations/env.py]: Codacy Issue found: [Invalid constant name "logger"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2611593900/issues/source?bid=3233864&fileBranchId=3299682#l14)

File: tests/test_google_oauth.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+            """For Google"""
+            self.assertTrue("You denied access" in self.app.get(
+                "/gCallback/?state=dummy_state&code=dummy_code&error=access denied").data)
+            self.assertTrue("Error encountered" in self.app.get(
+              "/gCallback/?state=dummy_state&code=dummy_code&error=12234").data)
+            self.assertTrue("/admin/login" in self.app.get("/gCallback/?no_code_and_state").data)
+            self.assertEqual(self.app.get("/gCallback/1234").status_code, 404)
+
+            """For Facebook"""
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [String statement has no effect](https://www.codacy.com/app/triplez/open-event-orga-server/file/2611715380/issues/source?bid=3233864&fileBranchId=3299682#l60)

File: tests/test_google_oauth.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+            directly to the admin page"""
+        with app.test_request_context():
+            register(self.app, 'test', 'email@gmail.com', 'test')
+            login(self.app, 'test', 'test')
+
+            """For 'Login with Google' """
+            self.assertTrue('/admin/' in self.app.get('/gCallback/?state=dummy_state&code=dummy_code)').data)
+            self.assertEqual(self.app.get('/gCallback/?state=dummy_state&code=dummy_code)').status_code, 302)
+
+            """For 'Login with Facebook' """
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [String statement has no effect](https://www.codacy.com/app/triplez/open-event-orga-server/file/2611715380/issues/source?bid=3233864&fileBranchId=3299682#l27)

File: tests/test_google_oauth.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+        self.app = Setup.create_app()
+
+    def test_user_already_logged_in(self):
+        """If the user is already logged in then on clicking 'Login with Google' he should be redirected
+            directly to the admin page"""
+        with app.test_request_context():
+            register(self.app, 'test', 'email@gmail.com', 'test')
+            login(self.app, 'test', 'test')
+
+            """For 'Login with Google' """
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [String statement has no effect](https://www.codacy.com/app/triplez/open-event-orga-server/file/2611715380/issues/source?bid=3233864&fileBranchId=3299682#l23)

File: tests/test_google_oauth.py [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+        """Tests whether the value of code is retained in the uri"""
+        client = WebApplicationClient('email@gmail.com')
+        uri = "https://gCallback/?state=dummy_state&code=dummy_code"
+        self.assertTrue('dummy_code' in client.parse_request_uri_response(uri, state='dummy_state').values())
+
+    def test_error_return(self):
+        """This tests the various errors returned by callback function"""
+        with app.test_request_context():
+
+            """For Google"""
```
Comment (juslee) [inline tests/test_google_oauth.py]: Codacy Issue found: [String statement has no effect](https://www.codacy.com/app/triplez/open-event-orga-server/file/2611715380/issues/source?bid=3233864&fileBranchId=3299682#l52)

**Comment by aditya1702** [comment]
@leto please review

**Comment by juslee** [comment]
@aditya1702 Move all secrets to environment variables and/or read from configuration files instead of hard coding.

**Comment by aditya1702** [comment]
Closing this because of another PR made for the gsoc branch, #478

**Pull request closed without merging**